nntrainer/tensor/blas_interface.cpp

```diff
@@ -235,22 +235,24 @@ static void scopy_int8_to_fp16(const unsigned int N, const uint8_t *X,
 }
 
 static void ewvm_FP16(const unsigned int N, const _FP16 *X, const _FP16 *Y,
-                      _FP16 *Z) {
+                      _FP16 *Z, const float alpha) {
 #if (defined USE__FP16 && USE_NEON)
```

Comment:
This preprocessor expression is not recommended.

https://stackoverflow.com/questions/1312132/boolean-in-ifdef-is-ifdef-a-b-the-same-as-if-defineda-definedb

Reply:
Then, as I presume:
`defined USE__FP16 && USE_NEON` to `defined USE__FP16 && define USE_NEON` ?

Reply:
what's the difference b/w `USE__FP16` and `USE_NEON` in the first place?

Reply:
First, my apologies for lack of understanding with those... but as far as I am concerned:

it is to differentiate fp16-enabled true/false with android/non-android build
neon is only valid for ARM.

Reply:
`USE_NEON` is currently enabled for both Android/non-android in `meson.build`. maybe we should fix this.

```
if get_option('enable-fp16')
   arch = target_machine.cpu_family()
   if get_option('platform') == 'android'
     add_project_arguments('-mfp16-format=ieee', language: ['c', 'cpp'])
     extra_defines += '-DENABLE_FP16=1'
     extra_defines += '-DUSE__FP16=1'
     extra_defines += '-DUSE_NEON=1'
   elif arch == 'aarch64'
     add_project_arguments('-mfp16-format=ieee', language: ['c', 'cpp'])
     extra_defines += '-DENABLE_FP16=1'
     extra_defines += '-DUSE__FP16=1'
     extra_defines += '-DUSE_NEON=1'
   else
     has_avx512fp16 = cc.has_argument('-mavx512fp16')
     if (has_avx512fp16)
       # add_project_arguments(['-mavx512fp16'], language: ['c','cpp'])
       message ('Float16 for x86_64 enabled. Modern gcc-x64 genrally supports float16 with _Float16. -mavx512fp16 added for hardware acceleration')
       extra_defines += '-DENABLE_FP16=1'
     else
       warning ('Float16 for x86_64 enabled. However, software emulation is applied for fp16, making it slower and inconsistent. Use GCC 12+ for AVX512 FP16 support. This build will probably fail unless you bring a compiler that supports fp16 for x64.')
     endif
   endif  
endif
```

Reply:
Also, I think we can replace `defined USE__FP16 && defined USE_NEON` to just `#if USE_NEON` since `USE__FP16` is inclusive

Reply:
First of all, `aarch64` also supports neon since it is 64bit ARM I guess?
Second of all, we still need it because neon also do work for float32.

For example,
1. fp16  / fp32 with no neon -> raw BLAS function or something else ( cBLAS / OpenCL, ... )
2. fp16 / fp32 with neon -> neon BLAS function

sometimes we might want to build with option such as: using neon but not fp16 (though it is not a typical case for now)

Reply:
I mean we are making more options like:
fp32-exclusive / fp16fp32 both  || neon / non-neon

Reply:
thanks for clarifying! to summarize, would things be like this?

- [x] USE_NEON=1 & USE__FP16=1: fp16/fp32 with neon BLAS
- [ ] (WTD) USE_NEON=1 & USE__FP16=0: fp32 only with neon BLAS 
- [ ] (WTD) USE_NEON=0 & USE__FP16=1: fp16/fp32 only with cBlas, openCL
- [ ] (WTD) USE_NEON=0 & USE__FP16=0: fp32 only with cBlas, openCL

Reply:
Happy to help :) That is correct.

Reply:
Change ```#if defined KEYWORD``` --> ```#if defined (KEYWORD)```.
And, for ```#if defined X && defined Y```, use ```if defined(X) && defined(Y)```, for the readability and compatibility with compilers.

Reply:
Thank you for your clear guide!
Applied suggested changes.